src/nrnoc/init.c

```diff
@@ -511,16 +498,24 @@ void nrn_register_mech_common(
 /*printf("%s %s\n", m[0], m[1]);*/
 	if (strcmp(m[0], "0") == 0) { /* valid by nature */
 	}else if (m[0][0] > '9') { /* must be 5.1 or before */
-fprintf(stderr, "Mechanism %s needs to be re-translated.\n\
+Fprintf(stderr, "Mechanism %s needs to be re-translated.\n\
 It's pre version 6.0 \"c\" code is incompatible with this neuron version.\n", m[0]);
-		nrn_exit(1);
+		if (nrn_load_dll_recover_error()) {
+			hoc_execerror("Mechanism needs to be retranslated:", m[0]);
+		}else{
+			nrn_exit(1);
+		}
 	}else if (strcmp(m[0], nmodl_version_) != 0){
-fprintf(stderr, "Mechanism %s needs to be re-translated.\n\
+Fprintf(stderr, "Mechanism %s needs to be re-translated.\n\
 It's version %s \"c\" code is incompatible with this neuron version.\n",
-			m[1], m[0]);
-		nrn_exit(1);
+m[1], m[0]);
```

Comment:
something for later - indentation needs to be fixed in the codebase to improve readability.